sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -318,6 +318,11 @@ object Cast extends QueryErrorsBase {
     case _ if from == to => true
     case (NullType, _) => true
     case (_: NumericType, _: NumericType) => true
+    case (_: AtomicType, CharType(_) | VarcharType(_))
+      if !SQLConf.get.preserveCharVarcharTypeInfo => false
+    case (CharType(a), CharType(b)) => a <= b
+    case (CharType(a), VarcharType(b)) => a <= b
+    case (VarcharType(a), VarcharType(b)) => a <= b
     case (_: AtomicType, CharType(_) | VarcharType(_)) => false
```

Comment:
Ditto here, the code looks weird as well.
```suggestion
    case (CharType(a), CharType(b)) if SQLConf.get.preserveCharVarcharTypeInfo => a <= b
    case (CharType(a), VarcharType(b)) if SQLConf.get.preserveCharVarcharTypeInfo => a <= b
    case (VarcharType(a), VarcharType(b)) if SQLConf.get.preserveCharVarcharTypeInfo => a <= b
    case (_: AtomicType, CharType(_) | VarcharType(_)) => false
```